### PR TITLE
Implement generate dataset okta command

### LINKED ISSUE #399: Add command to generate dataset with okta as a source
### Is your feature request related to a specific problem?

This feature will allow any user to “discover” 3rd party SaaS products that their organization uses that may contain PII. 

This spec is for a feature that requires:

* a user to connect to centralized service that is aware of all the software organization employees, or user to connect to a centralized service that is aware of 3rd party services
* the output of the scan to be a list of 3rd party external SaaS providers that contain PII from the organization.

### Describe the solution you'd like

Command:
```
./venv/bin/fidesctl generate dataset okta 
```

Output file:
```
dataset:
- fides_key: salesforce_001
  organization_fides_key: default_organization
  name: public
  description: 'Fides Generated Description for Schema: salesforce initially found YYYY-MM-DD'
  meta: null
  data_categories: []
  data_qualifier: aggregated.anonymized.unlinked_pseudonymized.pseudonymized.identified
  collections: []

- fides_key: zendesk_001
  organization_fides_key: default_organization
  name: public
  description: 'Fides Generated Description for Schema: zendesk initially found YYYY-MM-DD'
  meta: null
  data_categories: []
  data_qualifier: aggregated.anonymized.unlinked_pseudonymized.pseudonymized.identified
  collections: []
```

### PR DESCRIPTION
**Pull request by earmenda**
Closes #399 

### Code Changes

* [x] Add okta as an optional connector in `setup.py`
* [x] Add `fidesctl_meta` field to Dataset model
* [x] Add `generate_dataset_okta` click command
* [x] Update `api_helpers.list_server_resources` to allow keys to filter response
* [x] Update `dataset.py` functions to match the generate/scan implementation in `system.py`. Gather all datasets using `get_all_server_datasets` instead of by `fides_key`. This will be important as we move away from fetching resources using this key. 
* [x] Add functions to fetch okta applications, transform into dataset model and write to manifest under `dataset.py`. Refactored to reuse methods where possible.
* [x] Add new environment variable to `pytest-external` make target and `Run external test suite` pr check

### Steps to Confirm

* [x] Manual testing of command `fidesctl generate dataset okta https://dev-78908748.okta.com manifest.yml`
* [x] Unit testing for transforming okta model
* [x] Integration testing fetch applications from developer account and writing manifest

### Pre-Merge Checklist

* [x] All CI Pipelines Succeeded
* [x] Documentation Updated
* [x] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created

### Description Of Changes

**Success Example**
```
export OKTA_CLIENT_TOKEN="shhhhh"
root@3fbeb87ea959:/fides/fidesctl# fidesctl generate dataset okta https://dev-78908748.okta.com manifest.yml
Generated dataset manifest written to manifest.yml
```
manifest.yml:
```
dataset:
- fides_key: 0oa4h45lj1tcpqU6W5d7
  organization_fides_key: default_organization
  name: saasure
  description: 'Fides Generated Description for Okta Application: Okta Admin Console'
  data_categories: []
  data_qualifier: aggregated.anonymized.unlinked_pseudonymized.pseudonymized.identified
  fidesctl_meta:
    resource_id: 0oa4h45lj1tcpqU6W5d7
  retention: No retention or erasure policy
  collections: []
- fides_key: 0oa4h45ln0xLKJnAw5d7
  organization_fides_key: default_organization
  name: okta_enduser
  description: 'Fides Generated Description for Okta Application: Okta Dashboard'
  data_categories: []
  data_qualifier: aggregated.anonymized.unlinked_pseudonymized.pseudonymized.identified
  fidesctl_meta:
    resource_id: 0oa4h45ln0xLKJnAw5d7
  retention: No retention or erasure policy
  collections: []
- fides_key: 0oa4h45lnodX7MHJB5d7
  organization_fides_key: default_organization
  name: okta_browser_plugin
  description: 'Fides Generated Description for Okta Application: Okta Browser Plugin'
  data_categories: []
  data_qualifier: aggregated.anonymized.unlinked_pseudonymized.pseudonymized.identified
  fidesctl_meta:
    resource_id: 0oa4h45lnodX7MHJB5d7
  retention: No retention or erasure policy
  collections: []
```

**Error Invalid Url**
```
root@3fbeb87ea959:/fides/fidesctl# fidesctl generate dataset okta https://dev-78908748.okta.casdasdasdsa manifest.yml
Failed to list Okta applications: Cannot connect to host dev-78908748.okta.casdasdasdsa:443 ssl:default [Name or service not known]
```

**Error Invalid Token**
```
root@3fbeb87ea959:/fides/fidesctl# fidesctl generate dataset okta https://dev-78908748.okta.com manifest.yml
Failed to list Okta applications: {'message': 'Okta HTTP 401 E0000011 Invalid token provided\n'}
```

**TODO:** I created an okta developer account for the fidesctl team under the email `fidesteam@ethyca.com`. This has a token which we will use for integration tests. @ThomasLaPiana will add the token to the github repo under the name `OKTA_FIDESCTL_CLIENT_TOKEN`. The token can be found in our 1pass account under `Okta Fidesctl Integration Test Token`

### CODE REVIEW

**Comment by ThomasLaPiana** [comment]
@earmenda done!

**Comment by earmenda** [comment]
Looks like the only PR check failing is the db scanning. This was just added in #429 @ThomasLaPiana any idea why this would fail now?

**Comment by earmenda** [comment]
@ThomasLaPiana 

This is happening because I made the logic no longer match datasets based on `fides_key.` So it is looking for a dataset with the name `public` since that's how it is generated. Instead we have a name of `fidesctl`.

```
dataset:
  - fides_key: public
    organization_fides_key: default_organization
    name: fidesctl
```
This related to #430 because we need to have a way of doing consistent lookups for match db datasets with ones in our manifest. 

I would like to change the lookup logic here to look at the new `fidesctl_meta` field for a `resource_id` but I don't have a good idea right now of what we would use for an unique identifier yet. I'm okay with changing the dataset name to `public` in manifest until we have a solution for #430.

**Comment by ThomasLaPiana** [comment]
@earmenda yeah lets definitely do that in another PR, I'm ok to change this to `public` to get this passing for now

**Comment by earmenda** [comment]
> @earmenda yeah lets definitely do that in another PR, I'm ok to change this to `public` to get this passing for now

Sounds good, done!

**Comment by earmenda** [comment]
Getting the expected error now:

```
The following fields are missing data category annotations:
	public.datasets.fidesctl_meta

Annotation coverage: 99%
```

Will fix it by annotating field

File: fidesctl/src/fidesctl/core/api_helpers.py
Comment (ThomasLaPiana) [inline fidesctl/src/fidesctl/core/api_helpers.py:85]: according to this it returns an empty list only if the resource is valid, but it looks like the code was updated to always return an empty list?

**Comment by ThomasLaPiana** [comment]
@earmenda looks like we still need to do docs updates (maybe just adding an example of an okta generate in the how-to section?) and adding the follow-up issue to get `fidesctl_meta` working for dataset scanning

**Review by ThomasLaPiana — CHANGES_REQUESTED** [review]
Code changes look great, but a few small tasks remain

File: fidesctl/src/fidesctl/core/api_helpers.py
Comment (earmenda) [reply fidesctl/src/fidesctl/core/api_helpers.py:85]: Yeahhh let me know what you think but I wanted to create a new issue on adding error handling to the api_helpers functions. Even though what the comment is saying is technically true, the logic is missing right now to actually do this properly. My proposal was to create an issue so that `get_server_resources` `get_server_resource` and `list_server_resources` raise errors when failures happen. The reason I didn't want to throw this in here is that I think there is some callers of these functions that would break if I added this. For instance in `relationships.py` we call `get_server_resources` without knowing the resource type so we need to make changes there to properly handle other return codes as we should.

**Comment by earmenda** [comment]
> @earmenda looks like we still need to do docs updates (maybe just adding an example of an okta generate in the how-to section?) and adding the follow-up issue to get `fidesctl_meta` working for dataset scanning

Will update docs now. Regarding `fidesctl_meta`, isnt that what this issue is for? https://github.com/ethyca/fides/issues/430

**Comment by ThomasLaPiana** [comment]
> > @earmenda looks like we still need to do docs updates (maybe just adding an example of an okta generate in the how-to section?) and adding the follow-up issue to get `fidesctl_meta` working for dataset scanning
> 
> Will update docs now. Regarding `fidesctl_meta`, isnt that what this issue is for? #430

so it is, I forgot I had created that already, it's been a long week or so I guess...

File: fidesctl/src/fidesctl/core/api_helpers.py
Comment (ThomasLaPiana) [inline fidesctl/src/fidesctl/core/api_helpers.py:85]:
```suggestion
    Returns an empty list if no resources are found or if the API returns an error.
```

**Review by ThomasLaPiana — APPROVED** [review]
LGTM! Ready to merge after the merge conflict is fixed

**Comment by conceptualshark** [comment]
👍 All good in the docs world.

**Pull request merged**
